### PR TITLE
issue/565/add_flex_cluster_ensemble

### PR DESCRIPTION
**Pull request by eduardojsbarroso**
Added flexibility to ClusterEnsemble. Now, the object `ClusterEnsemble` can be instantiated without a list of clusters and the user can call the function `make_individual_radial_profile()` for each cluster that they want to use.

### CODE REVIEW

**Review by m-aguena — APPROVED** [review]

**Comment by m-aguena** [comment]
@combet can you also take a look at this PR? As I made changes, it would be nice to have a second validation.
Thanks!

File: clmm/clusterensemble.py [hunk 1]
```diff
@@ -113,38 +110,69 @@ def make_individual_radial_profile(self, galaxycluster, bin_units, bins=10, erro
...
+            GalaxyCluster object with cluster metadata and background galaxy data
+        tan_component: string, optional
+            Name of the tangetial component binned column to be added in profile table.
+            Default: 'gt'
+        cross_component: string, optional
+            Name of the cross component binned profile column to be added in profile table.
+            Default: 'gx'
+        weights : str, None
+            Name of the weight column to be used in the added to the profile table.
+        """
```
Comment (combet) [inline clmm/clusterensemble.py:151]: - [ ] `profile_table` is missing from the docstrings
- [ ] Rephrase decription the the `weights` keyword. Currently there's a problem in the sentence

File: clmm/clusterensemble.py [hunk 2]
```diff
@@ -190,18 +225,21 @@ def compute_bootstrap_covariance(self, tan_component='gt', cross_component='gx',
         n_bootstrap : int
```
Comment (combet) [inline clmm/clusterensemble.py:225]: - [ ] `tan_component` and `cross_component` missing from the docstring

File: clmm/clusterensemble.py [hunk 3]
```diff
@@ -190,18 +225,21 @@ def compute_bootstrap_covariance(self, tan_component='gt', cross_component='gx',
...
             number of bootstrap resamplings
```
Comment (combet) [inline clmm/clusterensemble.py:226]: - [ ] `tan_component` and `cross_component` missing from the docstring

File: clmm/clusterensemble.py [hunk 4]
```diff
@@ -177,11 +207,16 @@ def compute_sample_covariance(self, tan_component='gt', cross_component='gx'):
         sample_cross_covariance : ndarray
             The sample covariance matrix for the stacked cross profile
         """
```
Comment (combet) [inline clmm/clusterensemble.py:209]: - [ ] `tan_component` and `cross_component` missing from the docstring

File: clmm/clusterensemble.py [hunk 5]
```diff
@@ -190,18 +225,21 @@ def compute_bootstrap_covariance(self, tan_component='gt', cross_component='gx',
...
                             [data_bootstrap[tan_component], data_bootstrap[cross_component]])
             gt_boot.append(gt), gx_boot.append(gx)
         n_catalogs = len(self.data)
         coeff = (n_catalogs/(n_catalogs-1))**2
-        self.bootstrap_tangential_covariance = coeff*np.cov(np.array(gt_boot).T, bias = False,ddof=0)
+        self.bootstrap_tangential_covariance = coeff*np.cov(np.array(gt_boot).T,
+                                                            bias = False, ddof=0)
         self.bootstrap_cross_covariance = coeff*np.cov(np.array(gx_boot).T, bias = False)
 
     def compute_jackknife_covariance(self, tan_component='gt', cross_component='gx', n_side=16):
```
Comment (combet) [inline clmm/clusterensemble.py:245]: [ ] `tan_component` and `cross_component` missing from the docstring

File: clmm/dataops/__init__.py [hunk 6]
```diff
@@ -570,13 +570,15 @@ def make_radial_profile(components, angsep, angsep_units, bin_units,
...
+        r_avg, comp_avg, comp_err, nsrc, binnumber, wts_sum = compute_radial_averages(
             source_seps, component, xbins=bins,
-            yerr=None if components_error is None else components_error[i], weights=weights)
+            yerr=None if components_error is None else components_error[i],
+            weights=weights)
         profile_table[f'p_{i}'] = comp_avg
         profile_table[f'p_{i}_err'] = comp_err
     profile_table['radius'] = r_avg
     profile_table['n_src'] = nsrc
+    profile_table['weights'] = wts_sum
```
Comment (combet) [inline clmm/dataops/__init__.py:581]: Wondering whether this should be called `profile_table['weight_sum']` for clarity?

File: clmm/utils.py [hunk 7]
```diff
@@ -153,6 +153,8 @@ def compute_radial_averages(xvals, yvals, xbins, yerr=None, error_model='ste', w
         Indices of the bins (corresponding to `xbins`) in which each value
         of `xvals` belongs.  Same length as `yvals`.  A binnumber of `i` means the
         corresponding value is between (xbins[i-1], xbins[i]).
+    wts_sum: bumpy.array
+        Sums of weights in each bin
```
Comment (combet) [inline clmm/utils.py:157]: Sums of weights in each bin --> Sum of individual galaxy weights in each radial bin?

**Review by combet — COMMENTED** [review]
Thanks @eduardojsbarroso and @m-aguena! 

Apart from some docstrings to fix/improve, all looks god to me.

File: clmm/utils.py (hunk 7 above)
Comment (m-aguena) [reply clmm/utils.py:157]: This is a generic statistical function, it does not concern galaxies specifically, but I made a minor update on this docstring.

File: clmm/dataops/__init__.py (hunk 6 above)
Comment (m-aguena) [reply clmm/dataops/__init__.py]: Good idea. Done!

File: clmm/utils.py (hunk 7 above)
Comment (combet) [reply clmm/utils.py]: Yeap, right!

**Review by combet — APPROVED** [review]
Thanks for the doctstring updates @m-aguena. All looks good now! Happy to approve the PR.

**Comment by m-aguena** [comment]
> @m-aguena - I think it could be useful to add `save` and `load` methods to the `ClusterEnsemble` class, similarly to what we have for the `GalaxyCluster` object. What do you think? If you agree, could be taken care of in this PR?

That is a great idea!

**Review by combet — APPROVED** [review]
Thanks for the new save/load functions @m-aguena. All looks good.

**Pull request merged**